cachito/workers/pkg_managers/pip.py

```diff
@@ -332,3 +335,296 @@ def _get_package_dirs(self):
+
+        setup(name="foo", version="1.0.0")
+
+    2) if __main__
+
+        import setuptools
+
+        name = "foo"
+        version = "1.0.0"
+
+        if __name__ == "__main__":
+            setuptools.setup(name=name, version=version)
+
+    3) my_setup()
+
+        import setuptools
+
+        def my_setup():
+            name = "foo"
+            version = "1.0.0"
+
+            setuptools.setup(name=name, version=version)
+
+        my_setup()
+
+    For examples 2) and 3), we do not actually resolve any conditions or check that the
+    function containing the setup() call is eventually executed. We simply assume that,
+    this being the setup.py script, setup() will end up being called no matter what.
+    """
+
+    def __init__(self, top_dir):
+        """
+        Initialize a SetupPY.
+
+        :param str top_dir: Path to root of project directory
+        """
+        super().__init__(top_dir, "setup.py")
+        self.__ast = NOTHING
+        self.__setup_branch = NOTHING
+
+    def get_name(self):
+        """
+        Attempt to extract package name from setup.py.
+
+        :rtype: str or None
+        """
+        name = self._get_setup_kwarg("name")
+        if not name or not isinstance(name, str):
+            # Invalid means empty string or not a string
+            log.debug("No name in setup.py, failed to resolve or invalid")
+            return None
+
+        log.debug("Found name in setup.py: %r", name)
+        return name
+
+    def get_version(self):
+        """
+        Attempt to extract package version from setup.py.
+
+        As of setuptools version 49.2.1, there is no special logic for passing
+        an iterable as version in setup.py. Unlike name, however, it does support
+        non-string arguments (except tuples with len() != 1, those break horribly).
+
+        https://github.com/pypa/setuptools/blob/5e60dc50e540a942aeb558aabe7d92ab7eb13d4b/setuptools/dist.py#L462
+
+        Rather than trying to keep edge cases consistent with setuptools, treat them
+        consistently within Cachito.
+
+        :rtype: str or None
+        """
+        version = self._get_setup_kwarg("version")
+        if not version:
+            # Invalid means any falsy value (0, None, "", ...)
+            log.debug("No version in setup.py, failed to resolve or invalid")
+            return None
+
+        version = any_to_version(version)
+        log.debug("Found version in setup.py: %r", version)
+        return version
+
+    @property
+    def _ast(self):
+        """
+        Try to parse AST if not already parsed.
+
+        Will not parse file (or try to) more than once.
+        """
+        if self.__ast is NOTHING:
+            log.debug("Parsing setup.py at %r", str(self._path))
+
+            try:
+                self.__ast = ast.parse(self._path.read_text(), self._path.name)
+            except SyntaxError as e:
+                log.debug("Syntax error when parsing setup.py: %s", e)
```

Comment:
I guess it should. In that case, taking into account your other log suggestion, I guess the following log messages could be `error` as well:

setup.cfg:
- [Failed to parse setup.cfg](https://github.com/chmeliik/cachito/blob/d55853ece62bb3f0b3f1767075895fe1b996e95b/cachito/workers/pkg_managers/pip.py#L178)
- [Version file not found](https://github.com/chmeliik/cachito/blob/d55853ece62bb3f0b3f1767075895fe1b996e95b/cachito/workers/pkg_managers/pip.py#L210)
- [Module not found](https://github.com/chmeliik/cachito/blob/d55853ece62bb3f0b3f1767075895fe1b996e95b/cachito/workers/pkg_managers/pip.py#L246)
- [Syntax error when parsing module](https://github.com/chmeliik/cachito/blob/d55853ece62bb3f0b3f1767075895fe1b996e95b/cachito/workers/pkg_managers/pip.py#L252)
- [Could not find attribute in module](https://github.com/chmeliik/cachito/blob/d55853ece62bb3f0b3f1767075895fe1b996e95b/cachito/workers/pkg_managers/pip.py#L260)

setup.py
- Failed to parse setup.py (this)
- [Setup call not found](https://github.com/chmeliik/cachito/blob/d55853ece62bb3f0b3f1767075895fe1b996e95b/cachito/workers/pkg_managers/pip.py#L510)
- [Setup kwarg is an unsupported expression](https://github.com/chmeliik/cachito/blob/d55853ece62bb3f0b3f1767075895fe1b996e95b/cachito/workers/pkg_managers/pip.py#L605)
- Variable cannot be resolved ([your other comment](https://github.com/release-engineering/cachito/pull/226#discussion_r467245837))

If you agree on these, I would make the change in a separate commit.
Another option would be to use `warning`, since none of these errors are considered fatal. What do you think?

Reply:
+1 to the separate commit idea.

`Warn` seems to be used to indicate potentially harmful situations, whereas `error` seems to be used to indicate the error events that might still allow the application to continue running. Both seem to be fitting for these scenarios. I will leave the best judgement to you while making the changes!

Reply:
You're right, both seem applicable here. I went with `error` in the end